futures-sink/src/channel_impls.rs

```diff
@@ -1,62 +1,66 @@
-use {Async, Sink, AsyncSink, StartSend, Poll};
+use {Async, Sink, Poll};
 use futures_core::task;
-use futures_channel::mpsc::{Sender, SendError, UnboundedSender};
-
-fn res_to_async_sink<T>(res: Result<(), T>) -> AsyncSink<T> {
-    match res {
-        Ok(()) => AsyncSink::Ready,
-        Err(x) => AsyncSink::Pending(x),
-    }
-}
+use futures_channel::mpsc::{Sender, ChannelClosed, UnboundedSender};
 
 impl<T> Sink for Sender<T> {
     type SinkItem = T;
-    type SinkError = SendError<T>;
+    type SinkError = ChannelClosed<T>;
 
-    fn start_send(&mut self, cx: &mut task::Context, msg: T) -> StartSend<T, SendError<T>> {
-        self.start_send(cx, msg).map(res_to_async_sink)
+    fn poll_ready(&mut self, cx: &mut task::Context) -> Poll<(), Self::SinkError> {
+        self.poll_ready(cx)
     }
 
-    fn flush(&mut self, _: &mut task::Context) -> Poll<(), SendError<T>> {
-        Ok(Async::Ready(()))
+    fn start_send(&mut self, msg: T) -> Result<(), Self::SinkError> {
+        self.start_send(msg)
     }
 
-    fn close(&mut self, _: &mut task::Context) -> Poll<(), SendError<T>> {
+    fn start_close(&mut self) -> Result<(), Self::SinkError> {
+        Ok(())
+    }
+
+    fn poll_flush(&mut self, _: &mut task::Context) -> Poll<(), Self::SinkError> {
         Ok(Async::Ready(()))
     }
 }
 
 impl<T> Sink for UnboundedSender<T> {
     type SinkItem = T;
-    type SinkError = SendError<T>;
+    type SinkError = ChannelClosed<T>;
 
-    fn start_send(&mut self, cx: &mut task::Context, msg: T) -> StartSend<T, SendError<T>> {
-        self.start_send(cx, msg).map(res_to_async_sink)
+    fn poll_ready(&mut self, cx: &mut task::Context) -> Poll<(), Self::SinkError> {
+        self.poll_ready(cx)
     }
 
-    fn flush(&mut self, _: &mut task::Context) -> Poll<(), SendError<T>> {
-        Ok(Async::Ready(()))
+    fn start_send(&mut self, msg: T) -> Result<(), Self::SinkError> {
+        self.start_send(msg)
     }
 
-    fn close(&mut self, _: &mut task::Context) -> Poll<(), SendError<T>> {
+    fn start_close(&mut self) -> Result<(), Self::SinkError> {
+        Ok(())
+    }
+
+    fn poll_flush(&mut self, _: &mut task::Context) -> Poll<(), Self::SinkError> {
         Ok(Async::Ready(()))
     }
 }
 
 impl<'a, T> Sink for &'a UnboundedSender<T> {
     type SinkItem = T;
-    type SinkError = SendError<T>;
+    type SinkError = ChannelClosed<T>;
+
+    fn poll_ready(&mut self, _: &mut task::Context) -> Poll<(), Self::SinkError> {
+        Ok(Async::Ready(()))
```

Comment:
Shouldn't this also check the inherent `poll_ready`? An unbounded channel can still be closed.

Reply:
If it's closed, it will throw an error in `start_send`. However, I think I agree with you that it would be good to allow the client to stop producing values in these cases.

Reply:
(note that to accomplish this, i'll be changing the inherent `poll_ready` from `&mut self` to `&self`)

Reply:
oh, it seems I can't actually do that, because `poll_ready` calls `poll_unparked` internally, which mutates `self.maybe_parked`.

Reply:
As a user, I find it surprising that passing `tx` would let `poll_ready` tell me it was closed, but `&tx` wouldn't.

Reply:
I don't disagree. Can you open an issue for this?

Reply:
So just to confirm, we're gonna tackle this in a follow-up PR?

Reply:
Yes, I think that's a good idea.